packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -1283,15 +1351,30 @@ class RenderEditable extends RenderBox {
     offset.applyContentDimensions(0.0, _maxScrollExtent);
   }
 
+  Offset _getPixelPerfectCursorOffset(Rect caretRect) {
+    final Offset caretPosition = localToGlobal(caretRect.topLeft);
+    final double pixelMultiple = 1.0 / _devicePixelRatio;
```

Comment:
these divisions are a bit hard to read. Why not just

```dart
final Offset logicalCaretPosition = localToGlobal(caretRect.topLeft);
final int physicalCaretX = (caretPosition.dx * _devicePixelRatio).round();
final int physicalCaretY = (caretPosition.dy * _devicePixelRatio).round();
final double pixelPerfectLogicalX = globalToLocal(physicalCaretX / _devicePixelRatio); // I'm not sure what the - caretPosition.dx was for
final double pixelPerfectLogicalY = globalToLocal(physicalCaretY / _devicePixelRatio);
```

Otherwise, aliasing around the edge seems like 'working as intended'. Just have your test not land on .499999999-.5.